controllers/configuration_controller.go

```diff
@@ -89,16 +95,33 @@ func (r *ConfigurationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 	klog.InfoS("reconciling Terraform Configuration...", "NamespacedName", req.NamespacedName)
 	applyJobName := configurationName + "-" + string(TerraformApply)
 
-	// Terraform destroy
 	if err := r.Get(ctx, req.NamespacedName, &configuration); err != nil {
-		klog.InfoS("performing Terraform Destroy", "Namespace", req.Namespace, "Name", req.Name)
 		if kerrors.IsNotFound(err) {
+			klog.ErrorS(err, "unable to fetch Configuration", "NamespacedName", req.NamespacedName)
+			err = nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if configuration.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !meta.FinalizerExists(&configuration, configurationFinalizer) {
+			meta.AddFinalizer(&configuration, configurationFinalizer)
+			return ctrl.Result{}, errors.Wrap(r.Client.Update(ctx, &configuration), configurationFinalizer)
+		}
+	} else {
+		// terraform destroy
+		if meta.FinalizerExists(&configuration, configurationFinalizer) {
+			klog.InfoS("performing Terraform Destroy", "Namespace", req.Namespace, "Name", req.Name)
 			destroyJobName := configurationName + "-" + string(TerraformDestroy)
 			klog.InfoS("Terraform destroy job", "Namespace", req.Namespace, "Name", destroyJobName)
-			err := terraformDestroy(ctx, r.Client, ns, req.Name, destroyJobName, tfInputConfigMapsName, applyJobName, r.ProviderName)
-			return ctrl.Result{}, errors.Wrap(err, "continue reconciling to destroy cloud resource")
+			err := terraformDestroy(ctx, r.Client, ns, configuration, req.Name, destroyJobName, tfInputConfigMapsName, applyJobName, r.ProviderName)
+			if err != nil {
+				return ctrl.Result{RequeueAfter: 3 * time.Second}, errors.Wrap(err, "continue reconciling to destroy cloud resource")
+			}
+			meta.RemoveFinalizer(&configuration, configurationFinalizer)
```

Comment:
What will happen if the removal failed?

Reply:
It will block until destroyJob is complete.
If destroyJob does not complete successfully, finalizer will not to be delete. So, configuration will always exist.